File: rust/src/medium/say.rs
pub fn encode(n: u64) -> String {
    let num_string = n.to_string();
    let length = num_string.len() - 1;

    let vec = num_string.chars().rev().collect::<Vec<_>>();
    let thousands = vec.chunks(3).rev().collect::<Vec<_>>();

    thousands
        .into_iter()
        .enumerate()
        .map(|(ind, chars)| {
            let chars = chars.iter().rev().copied().collect::<Vec<_>>();
            let power = 10u64.pow((length - ind) as u32);

            dbg!(&chars);
            match chars.as_slice() {
                [a, '0', '0'] => format!("{}{}", say_number(&[*a]), say_thousands(power)),
                [a, b, c] => format!(
                    "{}{} {}",
                    say_number(&[*a]),
                    say_thousands(power),
                    say_number(&[*b, *c])
                ),
                ['0'] => "zero".to_string(),
                _ => format!("{}{}", say_number(&chars), say_thousands(power)),
            }
        })
        .filter(|x| dbg!(x) != " " && x != "")
        .collect::<Vec<_>>()
        .join(" ")
}

fn say_number(chars: &[char]) -> String {
    match chars {
        ['1'] => "one".to_string(),
        ['2'] => "two".to_string(),
        ['3'] => "three".to_string(),
        ['4'] => "four".to_string(),
        ['5'] => "five".to_string(),
        ['6'] => "six".to_string(),
        ['7'] => "seven".to_string(),
        ['8'] => "eight".to_string(),
        ['9'] => "nine".to_string(),
        ['1', '0'] => "ten".to_string(),
        ['1', '1'] => "eleven".to_string(),
        ['1', '2'] => "twelve".to_string(),
        ['1', '3'] => "thirteen".to_string(),
        ['1', '4'] => "fourteen".to_string(),
        ['1', '5'] => "fifteen".to_string(),
        ['1', '6'] => "sixteen".to_string(),
        ['1', '7'] => "seventeen".to_string(),
        ['1', '8'] => "eighteen".to_string(),
        ['1', '9'] => "nineteen".to_string(),
        ['2', '0'] => "twenty".to_string(),
        ['2', n] => format!("twenty-{}", say_number(&[*n])),
        ['3', n] => format!("thirty-{}", say_number(&[*n])),
        ['4', n] => format!("forty-{}", say_number(&[*n])),
        ['5', n] => format!("fifty-{}", say_number(&[*n])),
        ['6', n] => format!("sixty-{}", say_number(&[*n])),
        ['7', n] => format!("seventy-{}", say_number(&[*n])),
        ['8', n] => format!("eighty-{}", say_number(&[*n])),
        ['9', n] => format!("ninety-{}", say_number(&[*n])),
        ['0', '0'] => "".to_string(),
        ['0', n] => say_number(&[*n]),
        _ => format!(""),
    }
}

fn say_thousands(count: u64) -> String {
    match count {
        100 => " hundred".to_string(),
        1000 => " thousand".to_string(),
        1000_000 => " million".to_string(),
        1000_000_000 => " billion".to_string(),
        1000_000_000_000 => " trillion".to_string(),
        _ => "".to_string(),
    }
}

#[test]
fn test_zero() {
    assert_eq!(encode(0), String::from("zero"));
}

#[test]
fn test_one() {
    assert_eq!(encode(1), String::from("one"));
}

#[test]
fn test_fourteen() {
    assert_eq!(encode(14), String::from("fourteen"));
}

#[test]
fn test_twenty() {
    assert_eq!(encode(20), String::from("twenty"));
}

#[test]
fn test_twenty_two() {
    assert_eq!(encode(22), String::from("twenty-two"));
}

#[test]
fn test_one_hundred() {
    assert_eq!(encode(100), String::from("one hundred"));
}

#[test]
fn test_one_thousand_two_hundred_thirty_four() {
    assert_eq!(
        encode(1234),
        String::from("one thousand two hundred thirty-four")
    );
}

#[test]
fn test_one_million_two() {
    assert_eq!(encode(1_000_002), String::from("one million two"));
}

File: elixir/lib/easy/say.ex
defmodule Say do
  @upper_limit 999_999_999_999

  @doc """
  Translate a positive integer into English.
  """
  @spec in_english(integer) :: {atom, String.t()}
  def in_english(0), do: {:ok, "zero"}

  def in_english(number) when number < 0 or number > @upper_limit,
    do: {:error, "number is out of range"}

  def in_english(number) do
    thousands =
      number |> Integer.digits() |> Enum.reverse() |> Enum.chunk_every(3) |> Enum.reverse()

    len = Enum.count(thousands)

    {:ok,
     thousands
     |> Enum.with_index()
     |> Enum.map(fn {digits, index} ->
       digits = Enum.reverse(digits)
       power = 1000 ** (len - index - 1)

       case digits do
         [0, 0, 0] -> ""
         [a, 0, 0] -> "#{say_number([a])}#{say_thousands(power * 100)}"
         [0, 0, c] -> "#{say_number([c])}#{say_thousands(power)}"
         [0, b, c] -> "#{say_number([b, c])}#{say_thousands(power)}"
         [a, b, c] -> "#{say_number([a])} hundred #{say_number([b, c])}#{say_thousands(power)}"
         _ -> "#{say_number(digits)}#{say_thousands(power)}"
       end
     end)
     |> Enum.reject(&(&1 == ""))
     |> Enum.join(" ")}
  end

  defp say_number([tens, ones]) when tens in 2..9 do
    tens =
      case tens do
        2 -> "twenty"
        3 -> "thirty"
        4 -> "forty"
        5 -> "fifty"
        6 -> "sixty"
        7 -> "seventy"
        8 -> "eighty"
        9 -> "ninety"
      end

    [tens, say_number([ones])]
    |> Enum.reject(&is_nil/1)
    |> Enum.join("-")
  end

  defp say_number(digits) do
    case digits do
      [1] -> "one"
      [2] -> "two"
      [3] -> "three"
      [4] -> "four"
      [5] -> "five"
      [6] -> "six"
      [7] -> "seven"
      [8] -> "eight"
      [9] -> "nine"
      [1, 0] -> "ten"
      [1, 1] -> "eleven"
      [1, 2] -> "twelve"
      [1, 3] -> "thirteen"
      [1, 4] -> "fourteen"
      [1, 5] -> "fifteen"
      [1, 6] -> "sixteen"
      [1, 7] -> "seventeen"
      [1, 8] -> "eighteen"
      [1, 9] -> "nineteen"
      [0, 0] -> nil
      [0, n] -> say_number(n)
      _ -> nil
    end
  end

  defp say_thousands(number) do
    case number do
      100 -> " hundred"
      1_000 -> " thousand"
      1_000_000 -> " million"
      1_000_000_000 -> " billion"
      _ -> nil
    end
  end
end
